pkg/resolver/related.go

```diff
@@ -22,3 +32,327 @@ type SearchRelatedResult struct {
+
+	// Build the relations query
+	s.buildRelationsQuery()
+
+	relations, relQueryError := s.pool.Query(context.TODO(), s.query, s.params...) // how to deal with defaults.
+	if relQueryError != nil {
+		klog.Errorf("Error while executing getRelations query. Error :%s", relQueryError.Error())
+		return relatedSearch
+	}
+
+	defer relations.Close()
+
+	// iterating through resulting rows and scaning data, destid  and destkind
+	for relations.Next() {
+		var kind, uid string
+		var level int
+		relatedResultError := relations.Scan(&uid, &kind, &level)
+
+		if relatedResultError != nil {
+			klog.Errorf("Error %s retrieving rows for relationships:%s", relatedResultError.Error(), relations)
+			continue
+		}
+		s.updateKindMap(uid, kind, relatedMap) // Store result in a map
+	}
+
+	// retrieve provided kind filters only
+	if len(s.input.RelatedKinds) > 0 {
+		s.relatedKindUIDs(relatedMap) // get uids for provided kinds
+		// Build Related kinds query
+		s.buildRelatedKindsQuery()
+		items, err := s.resolveItems() // Fetch the related kind items
+		if err != nil {
+			klog.Warning("Error resolving relatedKind items", err)
+		} else { // Convert to (kind, items) format
+			relatedSearch = s.searchRelatedResultKindItems(items)
+		}
+	} else { // Retrieve kind and count of related items
+		relatedSearch = s.searchRelatedResultKindCount(relatedMap)
+	}
+	klog.V(5).Info("relatedSearch: ", relatedSearch)
+	return relatedSearch
+}
+
+func (s *SearchResult) relatedKindUIDs(levelsMap map[string][]string) {
+	relatedKinds := pointerToStringArray(s.input.RelatedKinds)
+	s.uids = []*string{}
+	for _, kind := range relatedKinds {
+		s.uids = append(s.uids, stringArrayToPointer(levelsMap[kind])...)
+	}
+}
+
+func (s *SearchResult) searchRelatedResultKindCount(levelMap map[string][]string) []SearchRelatedResult {
+
+	relatedSearch := make([]SearchRelatedResult, len(levelMap))
+
+	i := 0
+	//iterating and sending values to relatedSearch
+	for kind, uidArray := range levelMap {
+		count := len(uidArray)
+		relatedSearch[i] = SearchRelatedResult{Kind: kind, Count: &count}
+		i++
+	}
+	return relatedSearch
+}
+
+func (s *SearchResult) searchRelatedResultKindItems(items []map[string]interface{}) []SearchRelatedResult {
+
+	relatedSearch := make([]SearchRelatedResult, len(s.input.RelatedKinds))
+	relatedItems := map[string][]map[string]interface{}{}
+
+	//iterating and sending values to relatedSearch
+	for _, currItem := range items {
+		currKind := currItem["kind"].(string)
+		kindItemList := relatedItems[currKind]
+
+		kindItemList = append(kindItemList, currItem)
+		relatedItems[currKind] = kindItemList
+	}
+
+	i := 0
+	//iterating and sending values to relatedSearch
+	for kind, items := range relatedItems {
+		relatedSearch[i] = SearchRelatedResult{Kind: kind, Items: items}
+		i++
+	}
+	return relatedSearch
+}
+
+func (s *SearchResult) updateKindMap(uid string, kind string, levelMap map[string][]string) {
+	s.mux.RLock() // Lock map to read
```

Comment:
I wonder if the lock is really needed here. As far as I can tell everything runs in the same go routine, I don't see `go func()` anywhere.

This won't cause a problem, just wondering if it could be simplified.

Reply:
Removed